src/rapids_singlecell/preprocessing/_normalize.py

```diff
@@ -98,14 +99,114 @@ def normalize_total(
             (128,),
             (X, X.shape[0], X.shape[1], int(target_sum)),
         )
+        return X
 
-    if inplace:
-        _set_obs_rep(adata, X, layer=layer)
 
-    if copy:
-        return adata
-    elif not inplace:
-        return X
+def _normalize_total_csr(X: sparse.csr_matrix, target_sum: int) -> sparse.csr_matrix:
+    from ._kernels._norm_kernel import _mul_csr
+
+    mul_kernel = _mul_csr(X.dtype)
+    mul_kernel(
+        (math.ceil(X.shape[0] / 128),),
+        (128,),
+        (X.indptr, X.data, X.shape[0], int(target_sum)),
+    )
+    return X
+
+
+def _normalize_total_dask(X: DaskArray, target_sum: int) -> DaskArray:
+    if isinstance(X._meta, sparse.csr_matrix):
+        from ._kernels._norm_kernel import _mul_csr
+
+        mul_kernel = _mul_csr(X.dtype)
+        mul_kernel.compile()
+
+        def __mul(X_part):
+            mul_kernel(
+                (math.ceil(X_part.shape[0] / 32),),
+                (32,),
+                (X_part.indptr, X_part.data, X_part.shape[0], int(target_sum)),
+            )
+            return X_part
+
+        X = X.map_blocks(__mul, meta=_meta_sparse(X.dtype))
+    elif isinstance(X._meta, cp.ndarray):
+        from ._kernels._norm_kernel import _mul_dense
+
+        mul_kernel = _mul_dense(X.dtype)
+        mul_kernel.compile()
+
+        def __mul(X_part):
+            mul_kernel(
+                (math.ceil(X_part.shape[0] / 128),),
+                (128,),
+                (X_part, X_part.shape[0], X_part.shape[1], int(target_sum)),
+            )
+            return X_part
+
+        X = X.map_blocks(__mul, meta=_meta_dense(X.dtype))
+    else:
+        raise ValueError(f"Cannot normalize {type(X)}")
+    return X
+
+
+def _get_target_sum(X: ArrayTypesDask) -> int:
+    if isinstance(X, sparse.csr_matrix):
+        return _get_target_sum_csr(X)
+    elif isinstance(X, DaskArray):
+        return _get_target_sum_dask(X)
+    else:
+        return cp.median(X.sum(axis=1))
+
+
+def _get_target_sum_csr(X: sparse.csr_matrix) -> int:
+    from ._kernels._norm_kernel import _get_sparse_sum_major
+
+    counts_per_cell = cp.zeros(X.shape[0], dtype=X.dtype)
+    sum_kernel = _get_sparse_sum_major(X.dtype)
+    sum_kernel(
+        (X.shape[0],),
+        (64,),
+        (X.indptr, X.data, counts_per_cell, X.shape[0]),
+    )
```

Comment:
Please refactor these lines into their own function that can then be applied to each block of the dask array and then do the masking + median at the end after either you do `map_blocks` or `_get_target_sum_csr`.  We do this "recursive" `map_blocks` in `scanpy` a lot and it works very well and makes things very easy to reason about.  These functions are basically identical.  Especially if compiling really costs nothing extra as you say, then this should be doable.  Maybe I'm missing something though.  This seems like a good point in favor of "add compile to some always-called function" as Phil was saying

Reply:
I see where you are coming from with this one. However I feel the compile might make this challenging. I have an idea on how to maybe fix this but this needs some testing and investigations. I could wrap the cuda kernel factory to call compile on the returned kernel. But I don't know if this than executed on the worker or host.